rfcs/0032-band-join.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: band_join
+authors:
+  - "Dylan Chen"
+start_date: "2022/12/26"
+---
+
+# Band Join
+
+A band join is a special type of non equi join in which key values in one data set must fall within the specified range (“band”) of the second data set. There is another concept, range join which is similar to the band join but more flexible. Unluckily, range join is too general for streaming queries. We can only implement range join for batch queries.
+
+```sql
+-- This is a band-join.
+select * from A join B on A.p between B.d - 10 and B.d + 20.
+
+-- This is not a band-join, but a range join.
+select * from A join B on A.p between B.start and B.end.
+```
+
+The band join has a nice property. The range condition of band join only involves 2 columns: one from the LHS the other from RHS, so we can always reverse the condition. For example, `A.p between B.d - 10 and B.d + 20` can be converted into `B.d between A.p - 20 and A.p + 10`. This is a crucial property for streaming queries, as we need to treat both sides of the join logically equivalent (both sides need to be built and probed).
```

Comment:
> This is a crucial property for streaming queries, as we need to treat both sides of the join logically equivalent (both sides need to be built and probed). 


Why do we need to probe both sides? Why is this different in streaming vs. batch?

Reply:
For streaming, updates from either left side or right side need to be reflected on the join output. Think about `StreamHashJoin`, we need to build a hash table for both sides and if an update comes from one side, we need to use it to probe the other side to find the matched rows. This is also applicable to the stream-band-join. As you can see, we need to  build some index structures for both sides in streaming, while in batch we just need to build an index structure for one side (the build side), because the input size is bounded in batch.